File: src/haskell/2020/Day23.hs
{-|
Module      : Day23
Description : Day 23: Crab Cups

<https://adventofcode.com/2020/day/23>
-}

module Main where

import Advent

import Control.Applicative (liftA2)
import Data.Bifunctor (second)
import Data.List (maximum, sort)

parseInput :: String -> [Int]
parseInput = map (read . pure)

destCup :: Int -> [Int] -> Int
destCup i ixs =
    case takeWhile (< i) $ sort ixs of
        [] -> maximum ixs
        a -> maximum a

playTurn :: (Int, [Int]) -> (Int, [Int])
playTurn (curr, cups) =
    let
        (b,a) = second (drop 1) $ span (/= curr) cups
        r = drop 3 a ++ b
        d = destCup curr r
    in
        ( head (takeWhile (/= d) r ++ [d])
        , [curr] ++ takeWhile (/= d) r ++ [d] ++ take 3 a ++ drop 1 (dropWhile (/= d) r)
        )

formatResult :: Int -> [Int] -> Int
formatResult r = read . foldl1 (++) . map show . take r . drop 1 . dropWhile (/= 1) . cycle

puzzle1 :: [Int] -> Int
puzzle1 = formatResult 8 . snd . (!! 100) . iterate playTurn . liftA2 (,) head id

main = do
    input <- parsedInput (2020, 23) parseInput
    print $ puzzle1 input

File: src/python/2020/day23.py
# Advent of Code 2020, day 23: Crab Cups
# https://adventofcode.com/2020/day/23

from helpers import read

data = [int(x) for x in read(2020, 23)]


def cup_map(cups, p2=False):
    cm = {cups[-1]: cups[0]}
    for i in range(len(cups) - 1):
        cm[cups[i]] = cups[i + 1]

    if p2:
        cm[1000000] = cups[0]
        cm[cups[-1]] = 10
        for x in range(10, 1000000):
            cm[x] = x + 1
    return cm


def play_cups(cups, runs, p2=False):
    cm = cup_map(cups, p2=p2)
    head = cups[0]

    for _ in range(runs):
        taken = [cm[head]]
        taken += [cm[taken[0]]]
        taken += [cm[taken[1]]]

        dest = head - 1
        while True:
            if dest not in taken and dest in cm:
                break
            dest -= 1
            if dest not in cm:
                if p2:
                    dest = 1000000
                else:
                    dest = 9

        old_dest = cm[dest]
        cm[head] = cm[taken[-1]]
        cm[dest] = taken[0]
        cm[taken[-1]] = old_dest

        head = cm[head]
    return cm


def p1(input):
    cm = play_cups(input, 100, p2=False)
    s = ""
    x = 1
    for i in range(8):
        s += str(cm[x])
        x = cm[x]
    print(s)


def p2(input):
    cm = play_cups(input, 10000000, p2=True)
    print(cm[1] * cm[cm[1]])


p1(data)
p2(data)
